readme.md
Charm
=====
Charm is a library for LÖVE that makes it easier to arrange and draw graphics, such as shapes, images, and text. It's great for managing complex layouts with specific alignment and grouping needs.

Example
-------
![](https://i.imgur.com/gPAE6Wa.gif)

```lua
function love.draw()
	layout
		:new 'rectangle'
			:beginChildren()
				:new 'transform'
					:beginChildren()
						:new('rectangle', 0, 0, 300, 250)
							:beginChildren()
								:new('paragraph', font, text, 280, 'left', 10, 10 - textScrollY)
									:color(0, 0, 0)
								textHeight = layout:get('@current', 'height')
							layout:endChildren()
							:fillColor(.8, .8, .8)
							:clip()
					:endChildren()
					:scale(animationProgress)
					:angle((1 + animationProgress) * math.pi)
			:endChildren()
			:left(100):top(100)
		:draw()
end
```

> This is a snippet from the [dialog](https://github.com/tesselode/charm/blob/master/demos/dialog.lua) demo.

Installation
------------
To use Charm, place charm.lua in your project, and then `require` it in each file where you need to use it:

```lua
local charm = require 'charm' -- if your charm.lua is in the root directory
local charm = require 'path.to.charm' -- if it's in subfolders
```

Demos
-----
To see a demo, run love on the charm folder with the name of a demo as an argument:
```
love . dialog
love . menu
```

Documentation
-------------
### [Tutorial](https://tesselode.github.io/charm/tutorials/01-basic-usage.md.html) | [API Reference](https://tesselode.github.io/charm/)

Contributing
------------
Charm is in very early development. Feel free to ask questions, point out bugs, and even make some pull requests! If you use this library in a game, let me know how it goes.


changes.md
## v0.4 - in progress
This release brings back mouse events and is generally aimed at polishing the parts of Charm that worked well in previous versions.

Changes:
- The `preserve` table in element classes is replaced with `clearMode` for more customization over how fields are cleared
- Elements have internal state again, but with some changes:
	- Elements must be named *before* they're created
	- Elements that aren't named will be automatically assigned a name. This means that all elements can have internal state, even if the user didn't set a name.
- Re-added mouse event functionality
	- Elements now have a very simple "event" system where they can emit events, triggering functions that were registered to the event on that frame. This is mainly to provide a nicer syntax for checking for mouse events.
- Removed the `Transform` element, as it was buggy and confusing
- Consolidated the `Text` and `Paragraph` elements into one `Text` element, which can optionally have a horizontal limit and line wrapping. The new `Text` element is also more memory efficient than the old `Paragraph` element.
- For now, you can no longer get an element by name. The names are purely for tracking the state of elements between frames.
- Added the `drawDebug` for displaying useful debugging info about elements

## v0.3 - 2/9/20
This release is aimed at refocusing Charm on being a good library for drawing and arranging graphics.

Main changes:
- Removed support for mouse events
- Removed element state
- Added the transform element for applying arbitrary transformations to child elements, like rotation and scaling
- Added `Shape` and `Points` base classes to make it easier to write certain kinds of custom elements
- Added `Line` and `Polygon` elements
- Renamed `Ui` to `Layout`

## v0.2 - 12/6/19
This release is aimed at making Charm more extensible.

Main changes:
- The element types are now full fledged classes, and you can make your own classes that inherit from the built-in ones
- Much of the behavior of Charm now lives in the element classes. For example, `x`, `middle`, `wrap`, etc. are all functions on the base element class, and the `Ui` object just forwards function calls to the currently selected element
  - Instead of using `Ui.getX`, you use `Ui:get('@current', 'x')`, which calls the property getter named `x` that lives in the element class
  - Element classes now manage their own state, and they can add anything they want to the state table
- There's some new callbacks, like `onAddChild`

## v0.1 - 8/29/19
Commit arbitrarily chosen to be the "first release"


demos/dialog.lua
local charm = require 'charm'

local font = love.graphics.newFont(18)

local layout = charm.new()

local text = [[Good afternoon. The Lakitu Bros., here, reporting live from just outside the Princess's castle. Mario has just arrived on the scene, and we'll be filming the action live as he enters the castle and pursues the missing Power Stars. As seasoned cameramen, we'll be shooting from the recommended angle, but you can change the camera angle by pressing the camera up, camera right, camera down, and camera left buttons. If we can't adjust the view any further, we'll buzz. To take a look at the surroundings, stop and press the camera up button. Press A to resume play. Switch camera modes with the R Button. Signs along the way will review these instructions. For now, reporting live, this has been the Lakitu Bros.]]
local textScrollY = 0
local textHeight = 0

local targetAnimationProgress = 0
local animationProgress = 0

function love.update(dt)
	if love.keyboard.isDown 'up' then
		textScrollY = textScrollY - 300 * dt
	end
	if love.keyboard.isDown 'down' then
		textScrollY = textScrollY + 300 * dt
	end
	textScrollY = math.min(textScrollY, textHeight - 230)
	textScrollY = math.max(textScrollY, 0)

	animationProgress = animationProgress + (targetAnimationProgress - animationProgress) * 10 * dt
end

function love.keypressed(key)
	if key == 'space' then
		targetAnimationProgress = targetAnimationProgress == 0 and 1 or 0
	end
end

function love.draw()
	layout
		:new 'rectangle'
			:beginChildren()
				:new 'transform'
					:beginChildren()
						:new('rectangle', 0, 0, 300, 250)
							:beginChildren()
								:new('paragraph', font, text, 280, 'left', 10, 10 - textScrollY)
									:color(0, 0, 0)
								textHeight = layout:get('@current', 'height')
							layout:endChildren()
							:fillColor(.8, .8, .8)
							:clip()
					:endChildren()
					:scale(animationProgress)
					:angle((1 + animationProgress) * math.pi)
			:endChildren()
			:left(100):top(100)
		:draw()
end
